From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Tue, 13 Oct 2026 00:02:55 -0700
Subject: [PATCH] jscript: Array.prototype.reduceRight

ECMAScript 5 has reduceRight beside reduce; it was missing. It folds from the
last element to the first.

diff --git a/dlls/jscript/array.c b/dlls/jscript/array.c
index dfd9eda..1424f99 100644
--- a/dlls/jscript/array.c
+++ b/dlls/jscript/array.c
@@ -1465,6 +1465,79 @@ done:
     return hres;
 }
 
+static HRESULT Array_reduceRight(script_ctx_t *ctx, jsval_t vthis, WORD flags, unsigned argc, jsval_t *argv, jsval_t *r)
+{
+    jsval_t callback_args[4], acc, new_acc;
+    BOOL have_value = FALSE;
+    IDispatch *callback;
+    jsdisp_t *jsthis;
+    UINT32 length, k;
+    HRESULT hres;
+
+    TRACE("\n");
+
+    hres = get_length(ctx, vthis, &jsthis, &length);
+    if(FAILED(hres)) {
+        FIXME("Could not get length\n");
+        return hres;
+    }
+
+    /* Fixme check IsCallable */
+    if(!argc || !is_object_instance(argv[0])) {
+        FIXME("Invalid arg %s\n", debugstr_jsval(argc ? argv[0] : jsval_undefined()));
+        hres = E_INVALIDARG;
+        goto done;
+    }
+    callback = get_object(argv[0]);
+
+    if(argc > 1) {
+        have_value = TRUE;
+        hres = jsval_copy(argv[1], &acc);
+        if(FAILED(hres))
+            goto done;
+    }
+
+    for(k = length; k--; ) {
+        hres = jsdisp_get_idx(jsthis, k, &callback_args[1]);
+        if(hres == DISP_E_UNKNOWNNAME) {
+            hres = S_OK;
+            continue;
+        }
+        if(FAILED(hres))
+            break;
+
+        if(!have_value) {
+            have_value = TRUE;
+            acc = callback_args[1];
+            continue;
+        }
+
+        callback_args[0] = acc;
+        callback_args[2] = jsval_number(k);
+        callback_args[3] = jsval_obj(jsthis);
+        hres = disp_call_value(ctx, callback, jsval_undefined(), DISPATCH_METHOD, ARRAY_SIZE(callback_args), callback_args, &new_acc);
+        jsval_release(callback_args[1]);
+        if(FAILED(hres))
+            break;
+
+        jsval_release(acc);
+        acc = new_acc;
+    }
+
+    if(SUCCEEDED(hres) && !have_value) {
+        WARN("No array element\n");
+        hres = JS_E_INVALID_ACTION;
+    }
+
+    if(SUCCEEDED(hres) && r)
+        *r = acc;
+    else if(have_value)
+        jsval_release(acc);
+done:
+    jsdisp_release(jsthis);
+    return hres;
+}
+
 static HRESULT Array_some(script_ctx_t *ctx, jsval_t vthis, WORD flags, unsigned argc, jsval_t *argv,
         jsval_t *r)
 {
@@ -1623,6 +1696,9 @@ static const builtin_prop_t Array_props[] = {
     {L"pop",                   Array_pop,                  PROPF_METHOD},
     {L"push",                  Array_push,                 PROPF_METHOD|1},
     {L"reduce",                Array_reduce,               PROPF_METHOD|PROPF_ES5|1},
+#ifndef SG_MUTANT_JS_NOREDUCERIGHT
+    {L"reduceRight",           Array_reduceRight,          PROPF_METHOD|PROPF_ES5|1},
+#endif
     {L"reverse",               Array_reverse,              PROPF_METHOD},
     {L"shift",                 Array_shift,                PROPF_METHOD},
     {L"slice",                 Array_slice,                PROPF_METHOD|2},
